### DISCUSSION TITLE
strategy to prefetch part of data in getServerSideProps and then be able to show different data based on filter / search /etc.

### DISCUSSION THREAD

**Discussion by t-zander** [vercel/next.js#52109] [Help] [answered]
### Summary

I get a list of paginated products initially in `getServerSideProps`.
User can search for products, load more, additionally sort them by price.
When user does this, I would like to update the visible items. I basically have 2 strategies, but not sure which one is better 🤔 
1. put search and additional filters to query string as query params. Then do `router.push()` which will trigger `getServerSideProps` with new query params and in the end will give new data.
2. on search or change of sorting, etc. fire client side request and show data, which is returned from the backend. But then, I am not a fan of this, because it means I'd have props from `getServerSideProps` and data there + I'd have data which is stored in client side state, when I do client side request. So I'd have to have 2 data objects, defined which one of them to actually show, don't do this client side request on first page render (because data will come from `getServerSideProps`, etc...

Does it mean that in next.js you either stick to `getServerSideProps` and refetch data there by doing `router.push` or you d everything client side?

**Comment by icyJoseph** [comment, marked as answer]
That sounds a bit rushed. 

Though refreshing GSSP props is a good idea, https://www.joshwcomeau.com/nextjs/refreshing-server-side-props/, the second option is pretty valid too. 

The GSSP data can still be used as initial value for a state object. You can still use push with shallow to prevent GSSP re-trigger, and just update the state object.

```js
const [data, setData] = useState(propsFromGSSP)
```

So then `setData` handles filters, adding more etc... this does mean that you should not work with `propsFromGSSP` anymore, instead use `data`.

There's one caveat, if `_app` uses pageProps, it won't have updates made to `data`.

And yet another caveat, if `propsFromGSSP` does change, we probably would want to setData to that new value.

**Reply by t-zander** [reply]
It makes sense and answers it. Thanks 🙂 Well, it's not a problem, it's just that I was not sure if this is a good way. Usually, when I have to put smth from props into the state, I think twice if there is another way. But here it looks like a completely valid case

**Reply by icyJoseph** [reply]
For sure, you do good. If we allow a bit more critic, it does feel like, change in requirements, code ownership, or just a plain refactor, can turn this into a sour situation.

I'll complement my answer with yet another caveat, that if `propsFromGSSP` does change, we probably want to setData to that new object, since it is possible to, without unmounting the page, change the props coming in from GSSP, see: https://github.com/vercel/next.js/discussions/45077#discussioncomment-4737341

If there was a counter there, or if we had placed the props into state, then we'd be in deep...
